RDKBNETWOR-66 : Implement MAP-E in RdkWanManager

Krithiksha11 opened the pull request:
Reason for change: Receive the MAP-E options from DhcpManager and configure the tunnel interface.
Test Procedure: Verify functionality by passing DHCPv6 Option 94 and the ip6tnl tunnel interface is correctly configured and brought up.
Testing Done : Results are captured in RDKBNETWOR-66
Risks: None.

Krithiksha11 commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -318,6 +332,31 @@ static int wan_tearDownMapt()
+    char wan_interface[BUFLEN_64]={0};
+
+    CcspTraceInfo(("%s %d - stop MAP-E\n", __FUNCTION__, __LINE__));
+    if (syscfg_get(NULL, SYSCFG_WAN_INTERFACE, wan_interface, sizeof(wan_interface)) != ANSC_STATUS_SUCCESS)
```
Addressed

Krithiksha11 commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -2814,47 +2903,69 @@ static eWanState_t wan_transition_mapt_down(WanMgr_IfaceSM_Controller_t* pWanIfa
+                CcspTraceError(("%s %d Error resetting MAP-T configuration", __FUNCTION__, __LINE__));
+            }
+
+            /* Clear DHCPv4 client */
```
Moved outside if condition

Krithiksha11 commented on source/WanManager/wanmgr_net_utils.c:
```diff
@@ -2732,3 +2743,222 @@ int  WanManager_send_and_receive_rs(DML_VIRTUAL_IFACE * p_VirtIf)
+    inet_ntop(AF_INET6, &in6, tunnel_source, sizeof(tunnel_source));
+
+    //tunnel_source
+    ret = v_secure_system("ip -6 tunnel add %s mode ip4ip6 remote %s local %s encaplimit none dev %s", TUNNEL_NAME, br_ipv6_prefix, tunnel_source, "erouter0");
```
Addressed

Krithiksha11 commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -3725,6 +3855,15 @@ static eWanState_t wan_state_dual_stack_active(WanMgr_IfaceSM_Controller_t* pWan
     }
 #endif //FEATURE_MAPT
+
+#ifdef FEATURE_MAPE
```
Combined the conditions

Krithiksha11 commented on source/WanManager/wanmgr_net_utils.c:
```diff
@@ -114,6 +114,17 @@ extern int stop_dhcpv4_client (dhcp_params * params);
+#define BUFF_SIZE_64          64
+#define BUFF_SIZE_1024        1024
+#define IPV4_ADDR_LEN_IN_BITS 32
+#define BUF_LEN (10 * (sizeof(struct inotify_event) + 255 + 1))
```
Removed unused macros

Krithiksha11 commented on source/WanManager/wanmgr_net_utils.c:
```diff
@@ -2732,3 +2743,222 @@ int  WanManager_send_and_receive_rs(DML_VIRTUAL_IFACE * p_VirtIf)
 }
 
+#ifdef FEATURE_MAPE
+ANSC_STATUS WanManager_MAPEConfiguration(ipc_map_data_t *dhcp6cMAPEMsgBody)
```
Moved to new file as suggested

rdkcmf-jenkins commented:
b'## **Blackduck scan failure details**

**Summary**: 0 **violation**s, 0 files pending approval, 2 files pending identification.
* Protex Server Path: /home/blackduck/github/wan-manager/134/rdkcentral/RdkWanManager

* Commit: 5185d676a47c63da87e4c1a0e49e35acea0ec7a5

 **Report detail:** [gist](https://gist.github.com/rdkcmf-jenkins/c2d7bd67acc0d71aa2dcb85fbcc52482)'

mhughesacn commented:
Hi @Krithiksha11 : Please append a credit to NOTICE at top level, to match the new DT code:
```
Copyright 2025 Deutsche Telekom AG.
Licensed under the Apache License, Version 2.0
```

S-Parthiban-Selvaraj commented on config/RdkWanManager_v2.xml:
```diff
@@ -1437,5 +1437,335 @@
+           <name>Enable</name>
+           <type>boolean</type>
+           <syntax>bool</syntax>
+           <writable>true</writable>
```
I think it would be better to keep these DMLs read-only, as we're only using RBUS events or IPC methods to send lease details to the WAN Manager.

If these DMLs are writable, it opens up the possibility for someone to modify the configuration directly, which could conflict with the current design.

I understand that the BBF specification defines them as writable, but allowing write access here could introduce issues.

mhughesacn commented:
Hi @Krithiksha11 : I just noticed this: If you make any more changes to this PR, please will you fix the headers on the new files you created.  They say "this component's Licenses.txt file", but newer headers say "this component's LICENSE file".  We changed the license name a long time ago, but headers have been slow to catch up.  Many thanks!

Krithiksha11 commented:
> Hi @Krithiksha11 : I just noticed this: If you make any more changes to this PR, please will you fix the headers on the new files you created. They say "this component's Licenses.txt file", but newer headers say "this component's LICENSE file". We changed the license name a long time ago, but headers have been slow to catch up. Many thanks!

Hi @mhughesacn, 

I have addressed the requested change. Please review them.

Thank you!

Krithiksha11 commented on source/WanManager/wanmgr_mape.c:
```diff
@@ -0,0 +1,248 @@
+    syscfg_unset (NULL, "mape_psid");
+    syscfg_unset (NULL, "mape_psid_len");
+    syscfg_unset (NULL, "mape_ipv4_address");
+    syscfg_unset (NULL, "mape_ipv4_address");
```
Removed the duplicate call.

Krithiksha11 commented on source/WanManager/wanmgr_interface_sm.c:
```diff
@@ -2050,7 +2106,7 @@ static eWanState_t wan_transition_wan_validated(WanMgr_IfaceSM_Controller_t* pWa
     p_VirtIf->IP.Ipv6ConnectivityStatus = WAN_CONNECTIVITY_UP;
 
-    if(p_VirtIf->IP.SelectedMode == MAPT_MODE && p_VirtIf->IP.SelectedModeTimerStatus != EXPIRED)
+    if((p_VirtIf->IP.SelectedMode == MAPT_MODE || p_VirtIf->IP.SelectedMode == MAPE_MODE)&& p_VirtIf->IP.SelectedModeTimerStatus != EXPIRED)
```
Added space between ) &&
